google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java

```diff
@@ -199,66 +226,113 @@ private void setCredentials(Properties properties) throws IOException {
-      Preconditions.checkNotNull(
-          refreshToken,
-          "No refresh token specified under the key: %s",
-          ConfigPropertyKey.REFRESH_TOKEN);
+      if (refreshToken != null) {
+        userCredentialsBuilder.setRefreshToken(refreshToken);
+      }
+
       String clientId = ConfigPropertyKey.CLIENT_ID.getPropertyValue(properties);
-      Preconditions.checkNotNull(
-          clientId, "No client ID specified under the key: %s", ConfigPropertyKey.CLIENT_ID);
+      if (clientId != null) {
+        userCredentialsBuilder.setClientId(clientId);
+      }
+
       String clientSecret = ConfigPropertyKey.CLIENT_SECRET.getPropertyValue(properties);
-      Preconditions.checkNotNull(
-          clientSecret,
-          "No client secret specified under the key: %s",
-          ConfigPropertyKey.CLIENT_SECRET);
-
-      return UserCredentials.newBuilder()
-          .setClientId(clientId)
-          .setClientSecret(clientSecret)
-          .setRefreshToken(refreshToken)
-          .build();
+      if (clientSecret != null) {
+        userCredentialsBuilder.setClientSecret(clientSecret);
+      }
+
+      return userCredentialsBuilder;
     }
 
     /**
-     * Creates and returns a {@link Credentials} object based on the service account configuration
-     * keys.
+     * Configures the credentials builder using the specified properties for service account
+     * configuration keys.
+     *
+     * @return the modified, potentially <em>different</em> credentials builder
      */
-    private Credentials createServiceAccountCredentials(Properties properties) throws IOException {
+    private GoogleCredentials.Builder configureServiceAccountCredentials(
+        Properties properties, Optional<GoogleCredentials.Builder> optionalBuilder)
+        throws IOException {
+      ServiceAccountCredentials.Builder builder;
+      if (optionalBuilder.isPresent()) {
+        Preconditions.checkState(
+            optionalBuilder.get() instanceof ServiceAccountCredentials.Builder,
+            "Cannot apply service account credential settings. Builder is already partially"
+                + " configured for the installed app/web flow.");
+        builder = (ServiceAccountCredentials.Builder) optionalBuilder.get();
+      } else {
+        builder =
+            ServiceAccountCredentials.newBuilder()
+                .setScopes(Collections.singleton(GOOGLE_ADS_API_SCOPE));
+      }
+
       String serviceAccountSecretsPath =
           ConfigPropertyKey.SERVICE_ACCOUNT_SECRETS_PATH.getPropertyValue(properties);
-      Preconditions.checkNotNull(
-          serviceAccountSecretsPath,
-          "No service account secrets path specified under the key: %s",
-          ConfigPropertyKey.SERVICE_ACCOUNT_SECRETS_PATH);
+      if (serviceAccountSecretsPath != null) {
+        // Unfortunately, ServiceAccountCredentials.Builder does not provide a fromStream() method
```

Comment:
`fromStream()` populates at least 7 different attributes of the instance, as well as some other internals like the transport. I thought the safer choice here would be to let `fromStream()` do its thing, and then we can confidently copy the user and scopes since that covers everything we do with the `ServiceCredentials` interface without worrying about what's happening internally as a result of those calls.

I worry that if we dig into `fromStream()` and make sure we're copying everything it sets, this'll easily break if that implementation changes.

Reply:
Good call, thanks for the info.